src/Compilers/Core/CommandLine/BuildClient.cs

```diff
@@ -149,36 +136,65 @@ public Task<RunCompilationResult> RunCompilationAsync(IEnumerable<string> origin
+        /// Runs the provided compilation on the server.  If the compilation cannot be completed on the server then null
+        /// will be returned.
+        /// </summary>
+        internal RunCompilationResult? RunServerCompilation(TextWriter textWriter, List<string> arguments, BuildPaths buildPaths, string libDirectory, string sessionName, string keepAlive)
         {
-            switch (response.Type)
+            BuildResponse buildResponse;
+
+            try
             {
-                case BuildResponse.ResponseType.MismatchedVersion:
-                    Console.Error.WriteLine(CommandLineParser.MismatchedVersionErrorText);
-                    return RunCompilationResult.Failed;
+                var buildResponseTask = RunServerCompilation(
+                    arguments,
+                    buildPaths,
+                    sessionName,
+                    keepAlive,
+                    libDirectory,
+                    CancellationToken.None);
+                buildResponse = buildResponseTask.Result;
+
+                Debug.Assert(buildResponse != null);
+                if (buildResponse == null)
+                {
+                    return null;
+                }
+            }
+            catch (Exception)
+            {
+                return null;
+            }
 
+            switch (buildResponse.Type)
+            {
                 case BuildResponse.ResponseType.Completed:
-                    var completedResponse = (CompletedBuildResponse)response;
-                    return ConsoleUtil.RunWithUtf8Output(completedResponse.Utf8Output, textWriter, tw =>
                     {
-                        tw.Write(completedResponse.Output);
-                        return new RunCompilationResult(completedResponse.ReturnCode, ranOnServer: true);
-                    });
+                        var completedResponse = (CompletedBuildResponse)buildResponse;
+                        return ConsoleUtil.RunWithUtf8Output(completedResponse.Utf8Output, textWriter, tw =>
+                        {
+                            tw.Write(completedResponse.Output);
+                            return new RunCompilationResult(completedResponse.ReturnCode, ranOnServer: true);
+                        });
+                    }
 
+                case BuildResponse.ResponseType.MismatchedVersion:
                 case BuildResponse.ResponseType.Rejected:
                 case BuildResponse.ResponseType.AnalyzerInconsistency:
-                    var exitCode = RunLocalCompilation(arguments, buildPaths, textWriter);
-                    return new RunCompilationResult(exitCode);
-
+                    // Build could not be completed on the server.
+                    return null;
                 default:
-                    throw new InvalidOperationException("Encountered unknown response type");
+                    // Will not happen with our server but hypothetically could be sent by a rouge server.  Should
```

Comment:
Why are the rouge servers any worse than regular servers? Got something against the color red?